File: app/src/main/java/com/mgaetan89/showsrage/adapter/EpisodesAdapter.java
package com.mgaetan89.showsrage.adapter;

import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.widget.PopupMenu;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.mgaetan89.showsrage.Constants;
import com.mgaetan89.showsrage.R;
import com.mgaetan89.showsrage.databinding.AdapterEpisodesListBinding;
import com.mgaetan89.showsrage.model.Episode;
import com.mgaetan89.showsrage.model.Show;
import com.mgaetan89.showsrage.presenter.EpisodePresenter;

import java.util.Collections;
import java.util.List;

public class EpisodesAdapter extends RecyclerView.Adapter<EpisodesAdapter.ViewHolder> {
	@NonNull
	private List<Episode> episodes = Collections.emptyList();

	private final boolean reversed;

	private final int seasonNumber;

	private final Show show;

	public EpisodesAdapter(@Nullable List<Episode> episodes, int seasonNumber, Show show, boolean reversed) {
		if (episodes == null) {
			this.episodes = Collections.emptyList();
		} else {
			this.episodes = episodes;
		}

		this.reversed = reversed;
		this.seasonNumber = seasonNumber;
		this.show = show;
	}

	@Override
	public int getItemCount() {
		return this.episodes.size();
	}

	public boolean isReversed() {
		return this.reversed;
	}

	@Override
	public void onBindViewHolder(ViewHolder holder, int position) {
		Episode episode = this.episodes.get(position);

		holder.bind(new EpisodePresenter(episode));

		if (holder.name != null) {
			holder.name.setText(holder.name.getResources().getString(R.string.episode_name, this.getEpisodeNumber(position), episode.getName()));
		}

		if (holder.status != null) {
			int status = episode.getStatusTranslationResource();
			String statusString = episode.getStatus();

			if (status != 0) {
				statusString = holder.status.getResources().getString(status);
			}

			holder.status.setText(statusString);
		}
	}

	@Override
	public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.adapter_episodes_list, parent, false);

		return new ViewHolder(view);
	}

	private int getEpisodeNumber(int position) {
		if (this.reversed) {
			return this.getItemCount() - position;
		}

		return position + 1;
	}

	public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener, PopupMenu.OnMenuItemClickListener {
		@Nullable
		public final TextView name;

		@Nullable
		public final TextView status;

		@Nullable
		private final ImageView actions;

		private final AdapterEpisodesListBinding binding;

		public ViewHolder(View view) {
			super(view);

			view.setOnClickListener(this);

			this.binding = DataBindingUtil.bind(view);

			this.actions = this.binding.includeContent.episodeActions;
			this.name = this.binding.includeContent.episodeName;
			this.status = this.binding.includeContent.episodeStatus;

			if (this.actions != null) {
				this.actions.setOnClickListener(this);
			}
		}

		public void bind(EpisodePresenter episode) {
			this.binding.setEpisode(episode);
		}

		@Override
		public void onClick(View view) {
			Context context = view.getContext();

			if (view.getId() == R.id.episode_actions) {
				PopupMenu popupMenu = new PopupMenu(context, this.actions);
				popupMenu.inflate(R.menu.episode_action);
				popupMenu.setOnMenuItemClickListener(this);
				popupMenu.show();
			} else {
				EpisodesAdapter adapter = EpisodesAdapter.this;
				Episode episode = adapter.episodes.get(this.getAdapterPosition());

				if (context != null) {
					Intent intent = new Intent(Constants.Intents.ACTION_EPISODE_SELECTED);
					intent.putExtra(Constants.Bundle.EPISODE_MODEL, episode);
					intent.putExtra(Constants.Bundle.EPISODE_NUMBER, adapter.getEpisodeNumber(this.getAdapterPosition()));
					intent.putExtra(Constants.Bundle.EPISODES_COUNT, adapter.getItemCount());
					intent.putExtra(Constants.Bundle.SEASON_NUMBER, adapter.seasonNumber);
					intent.putExtra(Constants.Bundle.SHOW_MODEL, adapter.show);

					LocalBroadcastManager.getInstance(context).sendBroadcast(intent);
				}
			}
		}

		@Override
		public boolean onMenuItemClick(MenuItem item) {
			if (this.actions != null) {
				Context context = this.actions.getContext();

				if (context != null) {
					EpisodesAdapter adapter = EpisodesAdapter.this;
					Intent intent = new Intent(Constants.Intents.ACTION_EPISODE_ACTION_SELECTED);
					intent.putExtra(Constants.Bundle.EPISODE_NUMBER, adapter.getEpisodeNumber(this.getAdapterPosition()));
					intent.putExtra(Constants.Bundle.INDEXER_ID, adapter.show.getIndexerId());
					intent.putExtra(Constants.Bundle.MENU_ID, item.getItemId());
					intent.putExtra(Constants.Bundle.SEASON_NUMBER, adapter.seasonNumber);

					LocalBroadcastManager.getInstance(context).sendBroadcast(intent);

					return true;
				}
			}

			return false;
		}
	}
}

File: app/src/main/kotlin/com/mgaetan89/showsrage/adapter/EpisodesAdapter.kt
package com.mgaetan89.showsrage.adapter

import android.content.Intent
import android.content.res.ColorStateList
import android.support.v4.content.ContextCompat
import android.support.v4.content.LocalBroadcastManager
import android.support.v4.view.ViewCompat
import android.support.v7.widget.PopupMenu
import android.support.v7.widget.RecyclerView
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import com.mgaetan89.showsrage.Constants
import com.mgaetan89.showsrage.R
import com.mgaetan89.showsrage.extension.inflate
import com.mgaetan89.showsrage.model.Episode
import com.mgaetan89.showsrage.presenter.EpisodePresenter
import io.realm.RealmRecyclerViewAdapter
import io.realm.RealmResults
import kotlinx.android.synthetic.main.adapter_episodes_list_content.view.episode_actions
import kotlinx.android.synthetic.main.adapter_episodes_list_content.view.episode_date
import kotlinx.android.synthetic.main.adapter_episodes_list_content.view.episode_name
import kotlinx.android.synthetic.main.adapter_episodes_list_content.view.episode_quality
import kotlinx.android.synthetic.main.adapter_episodes_list_content.view.episode_status

class EpisodesAdapter(episodes: RealmResults<Episode>, val seasonNumber: Int, val indexerId: Int, val reversed: Boolean) : RealmRecyclerViewAdapter<Episode, EpisodesAdapter.ViewHolder>(episodes, true) {
	override fun onBindViewHolder(holder: ViewHolder, position: Int) {
		val episode = this.getItem(position)?.takeIf { it.isValid } ?: return

		holder.bind(episode)
	}

	override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
		val view = parent.inflate(R.layout.adapter_episodes_list)

		return ViewHolder(view)
	}

	inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener, PopupMenu.OnMenuItemClickListener {
		private val actions = view.episode_actions
		private val date = view.episode_date
		private val name = view.episode_name
		private val quality = view.episode_quality
		private val status = view.episode_status

		init {
			this.actions.setOnClickListener(this)
			this.name.isSelected = true

			view.setOnClickListener(this)
		}

		fun bind(episode: Episode) {
			val context = this.itemView.context
			val presenter = EpisodePresenter(episode)

			this.date.text = presenter.getAirDate() ?: context.getString(R.string.never)
			this.name.text = context.getString(R.string.episode_name, this.getEpisodeNumber(this.adapterPosition), episode.name)
			this.quality.text = presenter.getQuality()

			val backgroundColor = ContextCompat.getColor(context, presenter.getStatusColor())
			val status = episode.getStatusTranslationResource()
			ViewCompat.setBackgroundTintList(this.status, ColorStateList.valueOf(backgroundColor))
			this.status.text = if (status != 0) context.getString(status) else episode.status
		}

		override fun onClick(view: View) {
			when (view.id) {
				R.id.episode_actions -> this.showActions()
				else -> this.notifyEpisodeSelected()
			}
		}

		override fun onMenuItemClick(item: MenuItem): Boolean {
			this.notifyEpisodeActionSelected(item)

			return true
		}

		private fun getEpisodeNumber(position: Int) = if (reversed) itemCount - position else position + 1

		private fun notifyEpisodeActionSelected(item: MenuItem) {
			Intent(Constants.Intents.ACTION_EPISODE_ACTION_SELECTED).also {
				it.putExtra(Constants.Bundle.EPISODE_NUMBER, this.getEpisodeNumber(this.adapterPosition))
				it.putExtra(Constants.Bundle.INDEXER_ID, indexerId)
				it.putExtra(Constants.Bundle.MENU_ID, item.itemId)
				it.putExtra(Constants.Bundle.SEASON_NUMBER, seasonNumber)

				LocalBroadcastManager.getInstance(this.itemView.context).sendBroadcast(it)
			}
		}

		private fun notifyEpisodeSelected() {
			if (this.adapterPosition in (0 until itemCount)) {
				val episode = getItem(this.adapterPosition)?.takeIf { it.isValid } ?: return

				Intent(Constants.Intents.ACTION_EPISODE_SELECTED).also {
					it.putExtra(Constants.Bundle.EPISODE_ID, episode.id)
					it.putExtra(Constants.Bundle.EPISODE_NUMBER, this.getEpisodeNumber(this.adapterPosition))
					it.putExtra(Constants.Bundle.EPISODES_COUNT, itemCount)
					it.putExtra(Constants.Bundle.INDEXER_ID, indexerId)
					it.putExtra(Constants.Bundle.SEASON_NUMBER, seasonNumber)

					LocalBroadcastManager.getInstance(this.itemView.context).sendBroadcast(it)
				}
			}
		}

		private fun showActions() {
			PopupMenu(this.actions.context, this.actions).also {
				it.inflate(R.menu.episode_action)
				it.setOnMenuItemClickListener(this)
				it.show()
			}
		}
	}
}
